src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -169,6 +169,15 @@ internal bool TryAddWithoutValidation(HeaderDescriptor descriptor, IEnumerable<s
         {
             ArgumentNullException.ThrowIfNull(values);
 
+            if (values is string?[] valuesArray)
```

Comment:
Is special-casing `string[]` meaningfully faster compared to just having an `IList<string>` path?
With dynamic PGO I'd be surprised if it's worth complicating the helper.

Reply:
Yes it's more faster, deleting the specialization for array[] produce the following result

| Method     | Job        | Toolchain                                                                                                      | data    | Mean     | Error    | StdDev   | Median   | Min      | Max      | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|----------- |----------- |--------------------------------------------------------------------------------------------------------------- |-------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| **AddHeaders** | **Job-UTSEIL** | pr | **Array**   | **58.54 ns** | **1.162 ns** | **1.142 ns** | **58.66 ns** | **56.63 ns** | **60.80 ns** |  **0.70** |    **0.02** | **0.0266** |     **336 B** |        **0.69** |
| AddHeaders | Job-SILKBC | main | Array   | 83.34 ns | 1.276 ns | 1.193 ns | 82.79 ns | 81.51 ns | 85.11 ns |  1.00 |    0.02 | 0.0388 |     488 B |        1.00 |
|            |            |                                                                                                                |         |          |          |          |          |          |          |       |         |        |           |             |

Reply:
Do you mind checking what that looks like if you use just 1 or 2 values instead?

Reply:
| Method    | Toolchain                                                                                                      | data               | Mean     | Error    | StdDev   | Median   | Min      | Max      | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|----------- |--------------------------------------------------------------------------------------------------------------- |------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| **AddHeaders** | pr | **ArrayWithOneValue**  | **23.84 ns** | **1.060 ns** | **1.178 ns** | **23.55 ns** | **22.73 ns** | **26.24 ns** |  **0.83** |    **0.04** | **0.0165** |     **208 B** |        **0.87** |
| AddHeaders | master      | ArrayWithOneValue  | 28.70 ns | 0.234 ns | 0.207 ns | 28.71 ns | 28.32 ns | 29.02 ns |  1.00 |    0.01 | 0.0190 |     240 B |        1.00 |
|                       |                                                                                                                |                    |          |          |          |          |          |          |       |         |        |           |             |
| **AddHeaders** |  pr | **ArrayWithTwoValues** | **37.22 ns** | **0.520 ns** | **0.486 ns** | **36.93 ns** | **36.77 ns** | **38.06 ns** |  **0.77** |    **0.07** | **0.0261** |     **328 B** |        **0.95** |
| AddHeaders | master     | ArrayWithTwoValues | 48.96 ns | 4.014 ns | 4.623 ns | 47.45 ns | 43.57 ns | 60.07 ns |  1.01 |    0.13 | 0.0274 |     344 B |        1.00 |

Reply:
Here are the result for 1/2 values :

* With special handling for array :

| Method     | Toolchain                                                                                                      | data               | Mean     | Error    | StdDev   | Median   | Min      | Max      | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|----------- |--------------------------------------------------------------------------------------------------------------- |------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| **AddHeaders** |  pr | **ArrayWithOneValue**  | **23.84 ns** | **1.060 ns** | **1.178 ns** | **23.55 ns** | **22.73 ns** | **26.24 ns** |  **0.83** |    **0.04** | **0.0165** |     **208 B** |        **0.87** |
| AddHeaders |  master      | ArrayWithOneValue  | 28.70 ns | 0.234 ns | 0.207 ns | 28.71 ns | 28.32 ns | 29.02 ns |  1.00 |    0.01 | 0.0190 |     240 B |        1.00 |
|            |            |                                                                                                                |                    |          |          |          |          |          |          |       |         |        |           |             |
| **AddHeaders** | pr | **ArrayWithTwoValues** | **37.22 ns** | **0.520 ns** | **0.486 ns** | **36.93 ns** | **36.77 ns** | **38.06 ns** |  **0.77** |    **0.07** | **0.0261** |     **328 B** |        **0.95** |
| AddHeaders | master      | ArrayWithTwoValues | 48.96 ns | 4.014 ns | 4.623 ns | 47.45 ns | 43.57 ns | 60.07 ns |  1.01 |    0.13 | 0.0274 |     344 B |        1.00 |

* Without special handling for array :

| Method     | Toolchain                                                                                                      | data               | Mean     | Error    | StdDev   | Median   | Min      | Max      | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|----------- |--------------------------------------------------------------------------------------------------------------- |------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| **AddHeaders** | pr | **ArrayWithOneValue**  | **26.95 ns** | **0.713 ns** | **0.821 ns** | **26.59 ns** | **26.01 ns** | **28.59 ns** |  **0.99** |    **0.03** | **0.0165** |     **208 B** |        **0.87** |
| AddHeaders | master      | ArrayWithOneValue  | 27.22 ns | 0.334 ns | 0.313 ns | 27.15 ns | 26.72 ns | 27.86 ns |  1.00 |    0.02 | 0.0190 |     240 B |        1.00 |
|            |            |                                                                                                                |                    |          |          |          |          |          |          |       |         |        |           |             |
| **AddHeaders** | pr | **ArrayWithTwoValues** | **50.85 ns** | **0.361 ns** | **0.320 ns** | **50.84 ns** | **50.36 ns** | **51.37 ns** |  **1.17** |    **0.03** | **0.0260** |     **328 B** |        **0.95** |
| AddHeaders | master      | ArrayWithTwoValues | 43.41 ns | 1.046 ns | 1.074 ns | 42.84 ns | 42.45 ns | 46.20 ns |  1.00 |    0.03 | 0.0273 |     344 B |        1.00 |